### PR TITLE
Constant addition as a separate step (MCOL-5541)

### PR DESCRIPTION
**Pull request by mariadb-AndreyPiskunov**
This patch detaches constant column processing logic from TNS into a separate step called TCS.

### CODE REVIEW

File: dbcon/joblist/orderby-tests.cpp [hunk 1]
```diff
@@ -29,6 +29,7 @@
 #include "funcexp.h"
 #include "jlf_common.h"
 #include "tupleannexstep.h"
+#include "tupleannexconstantstep.h"
```
Comment (drrtuy) [inline dbcon/joblist/orderby-tests.cpp]: You don't need this.

File: dbcon/joblist/jlf_tuplejoblist.cpp [hunk 2]
```diff
@@ -556,6 +558,47 @@ void adjustLastStep(JobStepVector& querySteps, DeliveredTableMap& deliverySteps,
...
 
+  //if (orderby_threads == 99)
+  //{
+    SJSTEP acs(new TupleAnnexConstantStep(jobInfo));
+
+   if (jobInfo.constantCol == CONST_COL_EXIST)
+      dynamic_cast<TupleAnnexConstantStep*>(acs.get())->addConstant(new TupleConstantStep(jobInfo));
+
+  if (jobInfo.hasDistinct)
+      dynamic_cast<TupleAnnexConstantStep*>(acs.get())->setDistinct();
```
Comment (drrtuy) [inline dbcon/joblist/jlf_tuplejoblist.cpp]: You are not supposed to do distinct in this step.

File: dbcon/joblist/jlf_tuplejoblist.cpp [hunk 3]
```diff
@@ -319,7 +321,7 @@ void updateExp2Cols(JobStepVector& expSteps, TableInfoMap& tableInfoMap, JobInfo
   }
 }
 
-void adjustLastStep(JobStepVector& querySteps, DeliveredTableMap& deliverySteps, JobInfo& jobInfo)
+void adjustLastStep(JobStepVector& querySteps, DeliveredTableMap& deliverySteps, JobInfo& jobInfo, const uint32_t orderby_threads)
```
Comment (drrtuy) [inline dbcon/joblist/jlf_tuplejoblist.cpp]: jobInfo.orderByThreads has this info so you don't need to change the API to propagate the value.

File: dbcon/joblist/jlf_tuplejoblist.cpp [hunk 4]
```diff
@@ -556,6 +558,47 @@ void adjustLastStep(JobStepVector& querySteps, DeliveredTableMap& deliverySteps,
...
+
+  TupleAnnexStep* as = dynamic_cast<TupleAnnexStep*>(deliverySteps[CNX_VTABLE_ID].get());
+  RowGroup rg2 = as->getDeliveredRowGroup();
+
+  if (jobInfo.trace)
+    cout << "Output RowGroup 2: " << rg2.toString() << endl;
+
+  AnyDataListSPtr spdlIn(new AnyDataList());
+  RowGroupDL* dlIn;
+  if (jobInfo.orderByColVec.size() > 0)
```
Comment (drrtuy) [inline dbcon/joblist/jlf_tuplejoblist.cpp]: This step doesn't sort so you don't need this if but you need an else value.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+    fRowGroupDeliver.serializeRGData(bs);
+  }
+
+  return rowCount;
+}
+
+void TupleAnnexConstantStep::execute()
+{
+  if (fDistinct)
+    executeWithDistinct();
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: The step doesn't distinct semantics so you don't need this if, executeWithDistinct and related stuff.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+      rgDataOut.reinit(fRowGroupOut, fRowGroupIn.getRowCount());
+      fRowGroupOut.setData(&rgDataOut);
+      fRowGroupOut.resetRowGroup(fRowGroupIn.getBaseRid());
+      fRowGroupOut.setDBRoot(fRowGroupIn.getDBRoot());
+      fRowGroupOut.getRow(0, &fRowOut);
+
+      for (uint64_t i = 0; i < fRowGroupIn.getRowCount() && !cancelled() && !fLimitHit; ++i)
+      {
+        // skip first limit-start rows
+        if (fRowsProcessed++ < fLimitStart)
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: This is for limit processing and this step doesn't have it.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 7]
```diff
@@ -0,0 +1,563 @@
...
+        {
+          fRowIn.nextRow();
+          continue;
+        }
+
+        if (UNLIKELY(fRowsReturned >= fLimitCount))
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: Same thing.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 8]
```diff
@@ -0,0 +1,563 @@
...
+        {
+          fLimitHit = true;
+          fJobList->abortOnLimit((JobStep*)this);
+          continue;
+        }
+
+        if (fConstant)
+          fConstant->fillInConstants(fRowIn, fRowOut);
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: This is the only action that must be in this loop.

**Comment by drrtuy** [comment]
You need to temporary disable fillInConstants in the original TNS code.

File: dbcon/joblist/orderby-tests.cpp (hunk 1 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/joblist/orderby-tests.cpp]: Removed.

File: dbcon/joblist/jlf_tuplejoblist.cpp (hunk 2 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/joblist/jlf_tuplejoblist.cpp]: Removed.

File: dbcon/joblist/jlf_tuplejoblist.cpp (hunk 3 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/joblist/jlf_tuplejoblist.cpp]: Removed.

File: dbcon/joblist/jlf_tuplejoblist.cpp (hunk 4 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/joblist/jlf_tuplejoblist.cpp]: Removed.

File: dbcon/joblist/tupleannexconstantstep.cpp (hunk 5 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/joblist/tupleannexconstantstep.cpp]: I changed it as we discussed.

File: dbcon/joblist/tupleannexconstantstep.cpp (hunk 6 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/joblist/tupleannexconstantstep.cpp]: Removed.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+#include "funcexp.h"
+#include "jobstep.h"
+#include "jlf_common.h"
+#include "tupleconstantstep.h"
+#include "limitedorderby.h"
+
+#include "tupleannexconstantstep.h"
+
+#define QUEUE_RESERVE_SIZE 100000
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: You don't need this.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+  fExtendedInfo = "TACS: ";
+  fQtc.stepParms().stepType = StepTeleStats::T_TACS;
+}
+
+TupleAnnexConstantStep::~TupleAnnexConstantStep()
+{
+  if (fConstant)
+    delete fConstant;
+
+  fConstant = NULL;
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]:
```suggestion
  fConstant = nullptr;
```

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+
+  try
+  {
+    bs.restart();
+
+    more = fOutputDL->next(fOutputIterator, &fRgDataOut);
+
+    if (more && !cancelled())
+    {
+      fRowGroupOut.setData(&fRgDataOut);
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: Could you add a method unserializeIntoData that returns a number of rows?

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+      fRowGroupOut.serializeRGData(bs);
+      rowCount = fRowGroupOut.getRowCount();
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: The rowCount should be an accumulator.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+    while (more)
+      more = fOutputDL->next(fOutputIterator, &fRgDataOut);
+
+    fEndOfResult = true;
+  }
+
+  if (fEndOfResult)
+  {
+    // send an empty / error band
+    fRgDataOut.reinit(fRowGroupOut, 0);
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: Could you pack these 3 calls: setData, resetRowGroup, reset into a RowGroup method and call it appropriately.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+      fRowGroupOut.setData(&fRgDataOut);
+      fRowGroupOut.serializeRGData(bs);
+      rowCount = fRowGroupOut.getRowCount();
+    }
+    else
+    {
+      while (more)
+        more = fOutputDL->next(fOutputIterator, &fRgDataOut);
+
+      fEndOfResult = true;
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: I am curious what are the cases when fEndOfResult is set to false.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+    StepTeleStats sts(fQueryUuid, fStepUuid, StepTeleStats::ST_START, 1);
+    postStepStartTele(sts);
+
+    while (more && !cancelled())
+    {
+      fRowGroupIn.setData(&rgDataIn);
+      fRowGroupIn.getRow(0, &fRowIn);
+      // Get a new output rowgroup for each input rowgroup to preserve the rids
+      rgDataOut.reinit(fRowGroupOut, fRowGroupIn.getRowCount());
+      fRowGroupOut.setData(&rgDataOut);
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: Reuse the method that I mentioned previously here.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+      fRowGroupOut.resetRowGroup(fRowGroupIn.getBaseRid());
+      fRowGroupOut.setDBRoot(fRowGroupIn.getDBRoot());
+      fRowGroupOut.getRow(0, &fRowOut);
+
+      for (uint64_t i = 0; i < fRowGroupIn.getRowCount() && !cancelled(); ++i)
+      {
+        fConstant->fillInConstants(fRowIn, fRowOut);
+        fRowGroupOut.incRowCount();
+
+        fRowsReturned++;
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: You can sum up the number of rows in the input RG only once.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+  catch (...)
+  {
+    handleException(std::current_exception(), logging::ERR_IN_PROCESS, logging::ERR_ALWAYS_CRITICAL,
+                    "TupleAnnexConstantStep::execute()");
+  }
+
+  while (more)
+    more = fInputDL->next(fInputIterator, &rgDataIn);
+
+  // Bug 3136, let mini stats to be formatted if traceOn.
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: This is obsolete so plz remove.

File: dbcon/joblist/tupleannexconstantstep.cpp [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+    oss << "    " << fConstant->toString();
+
+  oss << endl;
+
+  return oss.str();
+}
+
+void TupleAnnexConstantStep::printCalTrace()
+{
+  time_t t = time(0);
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.cpp]: Is there a way to optimize this time calculation?

File: dbcon/joblist/tupleannexconstantstep.h [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+  rowgroup::Row row1, row2;  // scratch space for distinct comparisons todo: make them private
+
+ protected:
+  void execute();
+  void formatMiniStats();
+  void printCalTrace();
+
+  // input/output rowgroup and row
+  rowgroup::RowGroup fRowGroupIn;
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexconstantstep.h]: Plz rename the attributes.

File: dbcon/joblist/tupleannexstep.cpp [hunk 20]
```diff
@@ -762,103 +718,53 @@ void TupleAnnexStep::finalizeParallelOrderByDistinct()
...
-  }    // end of if-else
+    }
+
+    if (fRowGroupOut.getRowCount() > 0)
+    {
+      fRowsReturned += fRowGroupOut.getRowCount();
+      fOutputDL->insert(rgDataOut);
+      rgDataOut.reinit(fRowGroupIn, rowgroup::rgCommonSize);
+      fRowGroupOut.setData(&rgDataOut);
+      fRowGroupOut.resetRowGroup(0);
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexstep.cpp]: You could reuse the mentioned method here.

File: utils/querytele/telestats.h [hunk 21]
```diff
@@ -111,6 +112,7 @@ struct StepTeleStats
     T_SQS,  // SubQueryStep
     T_TAS,  // TupleAggregateStep
     T_TNS,  // TupleAnnexStep
+    T_TACS,
```
Comment (drrtuy) [inline utils/querytele/telestats.h]: Add the comment here.

File: dbcon/joblist/jlf_tuplejoblist.cpp [hunk 22]
```diff
@@ -705,7 +736,8 @@ void addProjectStepsToBps(TableInfoMap::iterator& mit, BatchPrimitive* bps, JobI
     {
       //			if (jobInfo.trace && bps->tableOid() >= 3000)
       //				cout << "1 setting project BPP for " << tbps->toString() << " with "
-      //<< 					it->get()->toString() << " and " << (it+1)->get()->toString() << endl;
+      //<< 					it->get()->toString() << " and " << (it+1)->get()->toString()
```
Comment (drrtuy) [inline dbcon/joblist/jlf_tuplejoblist.cpp]: Plz remove these comments.

**Review by drrtuy — CHANGES_REQUESTED** [review]

File: dbcon/joblist/tupleannexstep.cpp [hunk 23]
```diff
@@ -256,11 +242,7 @@ void TupleAnnexStep::run()
       dlTimes.setFirstReadTime();
 
     // *DRRTUY Make this block conditional
```
Comment (drrtuy) [inline dbcon/joblist/tupleannexstep.cpp]: This is an outdated and misguiding comment so plz remove it.

File: dbcon/joblist/tupleconstantbooleanstep.cpp [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+
+    if (traceOn())
+    {
+      dlTimes.setFirstReadTime();
+      dlTimes.setLastReadTime();
+      dlTimes.setEndOfInputTime();
+      printCalTrace();
+    }
+
+    // Bug 3136, let mini stats to be formatted if traceOn.
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantbooleanstep.cpp]: I doubt the comment makes sense now.

File: dbcon/joblist/tupleconstantbooleanstep.cpp [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+      throw logic_error("No output data list for non-delivery constant step.");
+
+    outputDL_ = fOutputJobStepAssociation.outAt(0)->rowGroupDL();
+
+    if (outputDL_ == NULL)
+      throw logic_error("Output is not a RowGroup data list.");
+
+    if (traceOn())
+    {
+      dlTimes.setFirstReadTime();
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantbooleanstep.cpp]: I am positive this block can be squashed into a function but we can make it in the distinct step patch.

File: dbcon/joblist/tupleconstantonlystep.cpp [hunk 26]
```diff
@@ -0,0 +1,472 @@
...
+}
+
+const RowGroup& TupleConstantOnlyStep::getOutputRowGroup() const
+{
+  return rowGroupOut_;
+}
+
+void TupleConstantOnlyStep::setOutputRowGroup(const rowgroup::RowGroup& rg)
+{
+  throw runtime_error("Disabled, use initialize() to set output RowGroup.");
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantonlystep.cpp:73]: Is it a mandatory virtual method from a parent class?

Comment (drrtuy) [reply dbcon/joblist/tupleconstantonlystep.cpp:73]: Can't you just pass on it?

File: dbcon/joblist/tupleconstantonlystep.cpp [hunk 27]
```diff
@@ -0,0 +1,472 @@
...
+  pos.push_back(2);
+
+  deliverStringTableRowGroup(false);
+
+  for (uint64_t i = 0; i < jobInfo.deliveredCols.size(); i++)
+  {
+    const ConstantColumn* cc = dynamic_cast<const ConstantColumn*>(jobInfo.deliveredCols[i].get());
+
+    if (cc == NULL)
+      throw runtime_error("none constant column found.");
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantonlystep.cpp:114]: should be non-constant IMHO

File: dbcon/joblist/tupleconstantonlystep.cpp [hunk 28]
```diff
@@ -0,0 +1,472 @@
...
+  vector<uint32_t> pos;
+  pos.push_back(2);
+
+  deliverStringTableRowGroup(false);
+
+  for (uint64_t i = 0; i < jobInfo.deliveredCols.size(); i++)
+  {
+    const ConstantColumn* cc = dynamic_cast<const ConstantColumn*>(jobInfo.deliveredCols[i].get());
+
+    if (cc == NULL)
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantonlystep.cpp:113]: nullptr

File: dbcon/joblist/tupleconstantonlystep.cpp [hunk 29]
```diff
@@ -0,0 +1,472 @@
...
+    if (ct.colDataType == CalpontSystemCatalog::VARCHAR)
+      ct.colWidth++;
+
+    // Round colWidth up
+    if (ct.colWidth == 3)
+      ct.colWidth = 4;
+    else if (ct.colWidth == 5 || ct.colWidth == 6 || ct.colWidth == 7)
+      ct.colWidth = 8;
+
+    oids.push_back(-1);
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantonlystep.cpp:127]: You can construct these two vectors right away with a vec ctor with a default value.

File: dbcon/joblist/tupleconstantonlystep.cpp [hunk 30]
```diff
@@ -0,0 +1,472 @@
...
+
+void TupleConstantOnlyStep::constructContanstRow(const JobInfo& jobInfo)
+{
+  // construct a row with only the constant values
+  constRowData_.reset(new uint8_t[rowConst_.getSize()]);
+  rowConst_.setData(rowgroup::Row::Pointer(constRowData_.get()));
+  rowConst_.initToNull();  // make sure every col is init'd to something, because later we copy the whole row
+  const vector<CalpontSystemCatalog::ColDataType>& types = rowGroupOut_.getColTypes();
+
+  for (vector<uint64_t>::iterator i = indexConst_.begin(); i != indexConst_.end(); i++)
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantonlystep.cpp:154]: Use auto here.

File: dbcon/joblist/tupleconstantstep.h [hunk 31]
```diff
@@ -1,26 +1,25 @@
...
+*/
 
 #pragma once
 
+#include <queue>
+#include <boost/thread/thread.hpp>
+
 #include "jobstep.h"
-#include "threadnaming.h"
+#include "limitedorderby.h"
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantstep.h]: I don't think you need this header.

File: dbcon/joblist/tupleconstantstep.cpp [hunk 32]
```diff
@@ -1,27 +1,19 @@
...
+ of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ You should have received a copy of the GNU General Public License along with this program; if not, write to
+ the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+// #define NDEBUG
 #include <cassert>
 #include <sstream>
 #include <iomanip>
+#include <tr1/unordered_set>
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantstep.cpp]: Do you need this here?

File: dbcon/joblist/tupleconstantstep.cpp [hunk 33]
```diff
@@ -1,27 +1,19 @@
...
+#include <fstream>
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantstep.cpp]: debugging leftover

File: dbcon/joblist/tupleconstantstep.cpp [hunk 34]
```diff
@@ -34,53 +26,35 @@ using namespace messageqcpp;
...
+using namespace utils;
+
 #include "querytele.h"
 using namespace querytele;
 
+#include "funcexp.h"
+#include "jobstep.h"
 #include "jlf_common.h"
+#include "tupleconstantstep.h"
+#include "limitedorderby.h"
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantstep.cpp]: Plz check if you need this here.

File: dbcon/joblist/tupleconstantstep.cpp [hunk 35]
```diff
@@ -34,53 +26,35 @@ using namespace messageqcpp;
...
 #include "constantcolumn.h"
+#include "simplecolumn.h"
 using namespace execplan;
 
-#include "jobstep.h"
 #include "rowgroup.h"
 using namespace rowgroup;
 
+#include "hasher.h"
+#include "stlpoolallocator.h"
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantstep.cpp]: Plz check if you need this here

**Review by drrtuy — CHANGES_REQUESTED** [review]

File: dbcon/joblist/tupleconstantstep.cpp [hunk 36]
```diff
@@ -286,225 +259,173 @@ void TupleConstantStep::run()
...
+    rowGroupOut_.setStatus(status());
+    rowGroupOut_.serializeRGData(bs);
   }
 
   return rowCount;
 }
 
 void TupleConstantStep::execute()
 {
+  utils::setThreadName("TACS");
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantstep.cpp]: This is TCS now not TACS.

File: dbcon/joblist/tupleconstantstep.cpp [hunk 37]
```diff
@@ -572,273 +494,16 @@ void TupleConstantStep::printCalTrace()
 void TupleConstantStep::formatMiniStats()
 {
   ostringstream oss;
-  oss << "TCS "
-      << "UM "
+  oss << "TACS ";
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantstep.cpp]: ditto

**Review by drrtuy — APPROVED** [review]

**Pull request closed without merging**